File: backend-express/src/routes/safe-date.js
const router = require('express').Router();
const SafeDate = require('../models/SafeDate');
const User = require('../models/User');
const { auth } = require('../middleware/auth');
const { ok, created, err } = require('../utils/respond');

// ── POST /api/safe-date/start ─────────────────────────────────────────────────
router.post('/start', auth, async (req, res, next) => {
  try {
    // End any existing active session first
    await SafeDate.updateMany(
      { user: req.user._id, isActive: true },
      { isActive: false, endedAt: new Date() }
    );

    const { trustedContactIds = [], meetingWith = '', venue = '', expectedEndTime } = req.body;

    // Resolve trusted contact nicknames
    const contacts = [];
    if (trustedContactIds.length > 0) {
      const users = await User.find({ _id: { $in: trustedContactIds } })
        .select('nickname')
        .lean();
      users.forEach((u) => contacts.push({ userId: u._id, nickname: u.nickname }));
    }

    const session = await SafeDate.create({
      user: req.user._id,
      trustedContacts: contacts,
      meetingWith,
      venue,
      expectedEndTime: expectedEndTime ? new Date(expectedEndTime) : null,
      startedAt: new Date(),
    });

    created(res, session);
  } catch (e) {
    next(e);
  }
});

// ── POST /api/safe-date/update-location ──────────────────────────────────────
router.post('/update-location', auth, async (req, res, next) => {
  try {
    const { lat, lng } = req.body;
    if (lat == null || lng == null) return err(res, 'lat and lng required');

    const session = await SafeDate.findOneAndUpdate(
      { user: req.user._id, isActive: true },
      {
        'location.coordinates': [parseFloat(lng), parseFloat(lat)],
        lastCheckinAt: new Date(),
      },
      { new: true }
    );
    if (!session) return err(res, 'No active session', 404);
    ok(res, session);
  } catch (e) {
    next(e);
  }
});

// ── POST /api/safe-date/panic ─────────────────────────────────────────────────
router.post('/panic', auth, async (req, res, next) => {
  try {
    const { lat, lng } = req.body;

    const update = { panicTriggered: true, panicAt: new Date() };
    if (lat != null && lng != null) {
      update['location.coordinates'] = [parseFloat(lng), parseFloat(lat)];
    }

    const session = await SafeDate.findOneAndUpdate(
      { user: req.user._id, isActive: true },
      update,
      { new: true }
    );
    if (!session) return err(res, 'No active session', 404);

    // TODO: Send push notifications to trustedContacts via FCM here.
    // For now, trusted contacts can poll /api/safe-date/alerts.
    // In production: iterate session.trustedContacts, look up FCM tokens, fire alerts.

    ok(res, {
      ok: true,
      alerted: session.trustedContacts.length,
      session,
    });
  } catch (e) {
    next(e);
  }
});

// ── POST /api/safe-date/end ───────────────────────────────────────────────────
router.post('/end', auth, async (req, res, next) => {
  try {
    const session = await SafeDate.findOneAndUpdate(
      { user: req.user._id, isActive: true },
      { isActive: false, endedAt: new Date() },
      { new: true }
    );
    if (!session) return err(res, 'No active session', 404);
    ok(res, session);
  } catch (e) {
    next(e);
  }
});

// ── GET /api/safe-date/active ─────────────────────────────────────────────────
router.get('/active', auth, async (req, res, next) => {
  try {
    const session = await SafeDate.findOne({
      user: req.user._id,
      isActive: true,
    }).lean();
    ok(res, session);
  } catch (e) {
    next(e);
  }
});

// ── GET /api/safe-date/alerts ─────────────────────────────────────────────────
// Panic alerts where I am a trusted contact.
router.get('/alerts', auth, async (req, res, next) => {
  try {
    const alerts = await SafeDate.find({
      'trustedContacts.userId': req.user._id,
      panicTriggered: true,
    })
      .sort({ panicAt: -1 })
      .limit(20)
      .populate('user', 'nickname avatarUrl')
      .lean();

    ok(res, alerts);
  } catch (e) {
    next(e);
  }
});

module.exports = router;

File: app/lib/core/models/safe_date.dart
class SafeDateContact {
  final String userId;
  final String nickname;

  const SafeDateContact({required this.userId, required this.nickname});

  factory SafeDateContact.fromJson(Map<String, dynamic> json) {
    return SafeDateContact(
      userId: (json['userId'] ?? '').toString(),
      nickname: json['nickname'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {'userId': userId, 'nickname': nickname};
}

class SafeDate {
  final String id;
  final bool isActive;
  final DateTime startedAt;
  final DateTime? endedAt;
  final List<SafeDateContact> trustedContacts;
  final String meetingWith;
  final String venue;
  final DateTime? expectedEndTime;
  final bool panicTriggered;
  final DateTime? panicAt;
  final DateTime? lastCheckinAt;
  final double? lat;
  final double? lng;

  const SafeDate({
    required this.id,
    required this.isActive,
    required this.startedAt,
    this.endedAt,
    this.trustedContacts = const [],
    this.meetingWith = '',
    this.venue = '',
    this.expectedEndTime,
    this.panicTriggered = false,
    this.panicAt,
    this.lastCheckinAt,
    this.lat,
    this.lng,
  });

  factory SafeDate.fromJson(Map<String, dynamic> json) {
    final rawId = json['_id'] ?? json['id'] ?? '';
    final contacts = (json['trustedContacts'] as List? ?? [])
        .map((c) => SafeDateContact.fromJson(c as Map<String, dynamic>))
        .toList();

    // coordinates: [lng, lat] in GeoJSON
    double? lat, lng;
    final loc = json['location'];
    if (loc is Map) {
      final coords = loc['coordinates'] as List?;
      if (coords != null && coords.length >= 2) {
        lng = (coords[0] as num).toDouble();
        lat = (coords[1] as num).toDouble();
      }
    }

    return SafeDate(
      id: rawId.toString(),
      isActive: json['isActive'] as bool? ?? false,
      startedAt: json['startedAt'] != null
          ? DateTime.parse(json['startedAt'])
          : DateTime.now(),
      endedAt: json['endedAt'] != null
          ? DateTime.tryParse(json['endedAt'])
          : null,
      trustedContacts: contacts,
      meetingWith: json['meetingWith'] as String? ?? '',
      venue: json['venue'] as String? ?? '',
      expectedEndTime: json['expectedEndTime'] != null
          ? DateTime.tryParse(json['expectedEndTime'])
          : null,
      panicTriggered: json['panicTriggered'] as bool? ?? false,
      panicAt: json['panicAt'] != null
          ? DateTime.tryParse(json['panicAt'])
          : null,
      lastCheckinAt: json['lastCheckinAt'] != null
          ? DateTime.tryParse(json['lastCheckinAt'])
          : null,
      lat: lat,
      lng: lng,
    );
  }
}
